File: app/src/main/java/com/example/bookworm/Extension/DiffUtilCallback.java
package com.example.bookworm.Extension;

import androidx.recyclerview.widget.DiffUtil;

import com.example.bookworm.Feed.Comments.Comment;
import com.example.bookworm.Feed.items.Feed;
import com.example.bookworm.Core.UserData.UserInfo;

import java.util.ArrayList;

public class DiffUtilCallback extends DiffUtil.Callback {
    private ArrayList oldList;
    private ArrayList newList;

    public DiffUtilCallback(ArrayList Old, ArrayList New) {
        this.oldList = Old;
        this.newList = New;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Object oldItem = oldList.get(oldItemPosition);
        Object newItem = newList.get(newItemPosition);
        if (oldItem instanceof Comment && newItem instanceof Comment) {
            return ((Comment) oldItem).getCommentID() == ((Comment) newItem).getCommentID();
        } else if (oldItem instanceof Feed && newItem instanceof Feed) {
            return ((Feed) oldItem).getFeedID() == ((Feed) newItem).getFeedID();
        } else if (oldItem instanceof UserInfo && newItem instanceof UserInfo) {
            return ((UserInfo) oldItem).getToken() == ((UserInfo) newItem).getToken();
        }
        return oldItem == newItem;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {

        return oldList.get(oldItemPosition) == newList.get(newItemPosition);
    }
}

File: app/src/main/java/com/example/bookworm/extension/DiffUtilCallback.kt
package com.example.bookworm.extension

import androidx.recyclerview.widget.DiffUtil
import com.example.bookworm.core.userdata.UserInfo
import com.example.bookworm.bottomMenu.Feed.comments.Comment
import com.example.bookworm.bottomMenu.Feed.items.Feed
import java.util.ArrayList

class DiffUtilCallback(private val oldList: ArrayList<*>, private val newList: ArrayList<*>) :
    DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldList[oldItemPosition]
        val newItem = newList[newItemPosition]
        if (oldItem is Comment && newItem is Comment) {
            return oldItem.commentID === newItem.commentID
        } else if (oldItem is Feed && newItem is Feed) {
            return oldItem.feedID === newItem.feedID
        } else if (oldItem is UserInfo && newItem is UserInfo) {
            return oldItem.token === newItem.token
        }
        return oldItem === newItem
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition] ==newList[newItemPosition]
    }
}